---
title: "Single Cell V(D)J Analysis"
author: "UCD Bioinformatics Core"
output:
    html_document:
      keep_md: TRUE
---

Last Updated June 22, 2023

# Single Cell V(D)J Analysis

## Packages
```{r, warning=FALSE,error=FALSE,message=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE)){
    install.packages("BiocManager")
}

if (!any(rownames(installed.packages()) == "ggplot2")){
  BiocManager::install("ggplot2")
}

if (!any(rownames(installed.packages()) == "viridis")){
  BiocManager::install("virids")
}

if (!any(rownames(installed.packages()) == "knitr")){
  BiocManager::install("knitr")
}

if (!any(rownames(installed.packages()) == "kableExtra")){
  BiocManager::install("kableExtra")
}

if (!any(rownames(installed.packages()) == "dplyr")){
  BiocManager::install("dplyr")
}

if (!any(rownames(installed.packages()) == "tidyr")){
  BiocManager::install("tidyr")
}

if (!any(rownames(installed.packages()) == "magrittr")){
  BiocManager::install("magrittr")
}

if (!any(rownames(installed.packages()) == "scRepertoire")){
  BiocManager::install("scRepertoire")
}

if (!any(rownames(installed.packages()) == "circlize")){
  BiocManager::install("circlize")
}

library(ggplot2)
library(viridis)
library(tidyr)
library(magrittr)
library(knitr)
library(kableExtra)
library(dplyr)
library(scRepertoire)
library(Seurat)
library(circlize)

set.seed(1234) # arbitrary
```

## Download Cell Ranger results

The following code downloads the output from a cellranger vdj run that has been stored in the github repository. The system call does not work on all operating systems.

```{r, eval=FALSE}
options(timeout=1200)
download.file("https://raw.githubusercontent.com/ucdavis-bioinformatics-training/2023-June-Advanced-Topics-in-Single-Cell-RNA-Seq-VDJ/main/data_analysis/cellranger_vdj_results.zip", "cellranger_vdj_results.zip")
system("unzip cellranger_vdj_results.zip")
```

## Set-up
```{r}
experiment_name = "VDJ Example"
dataset_loc <- "./cellranger_vdj_results"
ids <- c("Pool1")
```

## Sequencing metrics
```{r}
metrics <- paste(dataset_loc, ids, "metrics_summary.csv", sep = "/")
metrics_table <- do.call("cbind", lapply(metrics, function(x) {
  as.data.frame(t(read.csv(x)))
  }))
colnames(metrics_table) <- ids
rownames(metrics_table) <- gsub(".", " ", rownames(metrics_table), fixed = TRUE)
metrics_table  %>%
  kable(caption = 'Cell Ranger Results') %>%
  pack_rows("Overview", 1, 3, label_row_css = "background-color: #666; color: #fff;") %>%
  pack_rows("Sequencing Characteristics", 4, 8, label_row_css = "background-color: #666; color: #fff;") %>%
  pack_rows("Mapping Characteristics", 9, 26, label_row_css = "background-color: #666; color: #fff;") %>%
  kable_styling("striped", fixed_thead = TRUE)
```

The majority of the following functions and figures come from [scRepertoire](https://ncborcherding.github.io/vignettes/vignette.html). We will be exploring and making changes to the code as we go, so please take notes and don't be afraid to experiment and ask questions!

## Read in Cell Ranger VDJ Data
```{r}
clonotypes <- paste(dataset_loc, ids, "filtered_contig_annotations.csv", sep = "/")
vdj <- combineTCR(lapply(clonotypes, read.csv),
                  samples = ids,
                  ID = ids,
                  cells = "T-AB",
                  removeMulti = TRUE)
# rename barcodes to match the format oligo-sample
vdj <- lapply(vdj, stripBarcode)
vdj <- lapply(vdj, function(df){
  df$barcode = paste(sapply(strsplit(df$barcode, split = "-"), "[[", 1), df$sample, sep = "-")
  df
})
# structure of the vdj object
class(vdj)
class(vdj[[1]])
head(vdj[[1]]) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover", "responsive"))
```

## Number of unique clonotypes
```{r}
quantContig(vdj, cloneCall="aa", group = "sample", scale = FALSE, exportTable = TRUE) %>%
  select(sample, contigs, total) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover", "responsive"))
```

## Distribution of clonotypes by abundance
```{r}
abundanceContig(vdj, cloneCall = "gene", group = "sample", scale = FALSE) +
  guides(color = "none")
abundanceContig(vdj, cloneCall = "gene", group = "sample", scale = FALSE, exportTable = TRUE) %>%
  arrange(desc(Abundance)) %>%
  filter(!is.na(CTgene)) %>%
  select(CTgene, Abundance) %>%
  slice(1:20) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover", "responsive"), fixed_thead = TRUE)
```

## Contig length distribution
```{r}
lengthContig(vdj, cloneCall="nt", scale=TRUE, chain = "both", group="sample") +
  guides(fill = "none")

vdj[[1]] %>%
  mutate(TRA = nchar(cdr3_aa1),
         TRB = nchar(cdr3_aa2)) %>%
  pivot_longer(cols = c(TRA, TRB),
               names_to = "chain",
               values_to = "chain_aa_length") %>%
  ggplot(aes(x = chain_aa_length)) +
  geom_histogram(binwidth = 1) +
  facet_wrap(~chain, ncol = 1) +
  theme_classic()

vdj[[1]] %>%
  mutate(TRA = nchar(cdr3_aa1),
         TRB = nchar(cdr3_aa2)) %>%
  pivot_longer(cols = c(TRA, TRB),
               names_to = "chain",
               values_to = "chain_aa_length") %>%
  ggplot(aes(x = chain_aa_length)) +
  geom_freqpoly(binwidth = 1) +
  facet_wrap(~chain, ncol = 1) +
  theme_classic()
```

### Gene usage

```{r}

```

### Relative abundance of clones by frequency
```{r}
clonalHomeostasis(vdj, cloneCall = "aa") +
  scale_fill_viridis_d(option = "plasma") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        axis.title.x = element_blank(),
        legend.title = element_blank())
```

### Relative abundance of clones by index

Clonal index 1 represents the most frequent clone in a given sample, while index 1000 represents the 1000th most frequent clone.
```{r}
clonalProportion(vdj, cloneCall = "aa", split = c(10, 50, 100, 500, 1000)) +
  scale_fill_viridis_d(option = "rocket", direction = -1) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        axis.title.x = element_blank(),
        legend.title = element_blank())
```

## TCR clustering

TCR clustering groups clonotypes together on sequence similarity (either nucleotide or amino acid, controlled by the "sequence" argument). This may be useful in order to identify cells that are producing receptors that may have similar properties, even though the clone differs.

A new column called "TRA_cluster" (or "TRB_cluster" depending on the chain used), is added to the output by the clustering function. Cluster names containing "LD" represent multi-clonal groups created by the function.

```{r, eval=FALSE}
tcr.clusters <- clusterTCR(vdj[[1]], chain = "TRA", sequence = "aa", threshold = 0.9)
# three most abundant aa sequences in three clusters
tcr.clusters[[1]] %>%
  count(cdr3_aa1, TRA_cluster) %>%
  filter(TRA_cluster %in% c("TRA:LD.31", "TRA:LD.23", "TRA:LD.55")) %>%
  group_by(TRA_cluster) %>%
  arrange(desc(n)) %>%
  slice(1:3)
# 5 largest clusters
tcr.clusters[[1]] %>%
  count(TRA_cluster) %>%
  filter(!is.na(TRA_cluster)) %>%
  arrange(desc(n)) %>%
  slice(1:5)
```

## Combine V(D)J and expression data

First, we need to download the expression data. The Seurat object located in the vdj_workshop directory on tadpole contains a subset of the data from the study, re-analyzed with different parameters.

```{bash, eval = FALSE}
scp hslyman@tadpole.genomecenter.ucdavis.edu:/share/workshop/vdj_workshop/R_objects/seurat_object.rds .
```

The the cell barcodes in the "barcode" column of the vdj object must match the barcodes in the expression object in order for the call to combineExpression to work. We handled this when we created the vdj object (using the stripBarcode function and a little bit of custom code), but will verify quickly that there are shared cell names. The "cloneTypes" argument allows the user to set the frequency values associated with each named frequency class.

```{r}
expression <- readRDS("seurat_object.rds")
# 3654 cell names in Seurat match cell names in scRepertoire
length(which(sapply(vdj, function(sample){ sample$barcode }) %in% Cells(expression)))
# add scRepertoire info to expression object
expression <- combineExpression(vdj, expression, cloneCall = "gene", chain = "both")
head(expression@meta.data) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover", "responsive"))
```
## Find markers for "large" clones
```{r}
DimPlot(expression, group.by = "cloneType") +
  scale_color_viridis_d(option = "plasma", direction = -1, end = 0.8)
Idents(expression) <- "cloneType"
large.markers <- FindMarkers(expression, ident.1 = "Large (0.01 < X <= 0.1)")
large.markers$gene <- rownames(large.markers)
view.markers <- large.markers %>%
  filter(p_val_adj < 0.05) %>%
  filter(!grepl("^TR[AB]", gene)) %>%
  slice(1:5)
# FeaturePlots
lapply(view.markers$gene, function(marker){
  FeaturePlot(expression,
              features = marker)
})  
# VlnPlots
lapply(view.markers$gene, function(marker){
  VlnPlot(expression,
          features = marker,
          group.by = "RNA_snn_res.0.25") +
    scale_fill_viridis_d(option = "turbo")
})
# view table
kable(view.markers) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "responsive"))
```

## Split V(D)J data into separate samples and re-run scRepertoire functions

Using the information contained in the Seurat object, we can now split our pooled VDJ data into multiple groups. 

```{r}
colnames(expression@meta.data)
```

Let's create a new vdj object from the Seurat object, grouping by sample.

```{r}
expression$Sample_Name_short <- gsub("batch1.", "", expression$Sample_Name)
vdj.sample <- expression2List(expression, split.by = "Sample_Name_short")
class(vdj.sample)
names(vdj.sample)
```

Then, all the same functions we ran above can be rerun.

```{r}
quantContig(vdj.sample, cloneCall = "aa", scale = FALSE) +
  scale_fill_viridis_d()
abundanceContig(vdj.sample, cloneCall = "gene", scale = FALSE) +
  scale_color_viridis_d()
clonalHomeostasis(vdj.sample, cloneCall = "aa") +
  scale_fill_viridis_d(option = "plasma") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        axis.title.x = element_blank(),
        legend.title = element_blank())
clonalProportion(vdj.sample, cloneCall = "aa", split = c(10, 50, 100, 500, 1000)) +
  scale_fill_viridis_d(option = "rocket", direction = -1) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        axis.title.x = element_blank(),
        legend.title = element_blank())
```

Some functions that we skipped over before because they are not very informative on a single sample are now worth running.

## Compare abundance of clonotypes across samples

```{r}
compareClonotypes(vdj.sample, numbers = 5, cloneCall = "aa", graph = "alluvial") +
  scale_fill_viridis_d(option = "turbo") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(caption = "Results of compareClonotypes() with numbers = 5.")
# list shared clones
compare.clones <- compareClonotypes(vdj.sample,
                                    numbers = 10000,
                                    cloneCall = "aa",
                                    exportTable = TRUE) %>%
  pivot_wider(names_from = Sample,
              values_from = Proportion,
              names_repair = "universal")
compare.clones$shared.by <- apply(compare.clones, 1, function(x){
  length(which(!sapply(x[2:5], is.na)))
})
# display 10 clones
compare.clones %>%
  filter(shared.by > 1) %>%
  arrange(desc(shared.by)) %>%
  select(-shared.by) %>%
  slice(1:10) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover", "responsive"))
```

### Calculate clonal overlap between samples

```{r}
clonalOverlap(vdj.sample, cloneCall = "aa", method = "overlap") +
  scale_fill_viridis(option = "mako") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```

### Calculate clonal diversity

```{r}
clonalDiversity(vdj.sample, cloneCall = "aa") +
  scale_fill_viridis_d() +
  theme(legend.title = element_blank())
```

### clonalOverlay

In our data set, I wasn't able to find a set of parameters to get the contour to plot using the clonalOverlay function as shown in the documentation [here](https://www.bioconductor.org/packages/release/bioc/vignettes/scRepertoire/inst/doc/vignette.html#62_clonalOverlay). If you figure out how to make it work, please let me know!

```{r}
# uses active identity
Idents(expression) <- expression$Sample_Name_short
clonalOverlay(expression,
              reduction = "umap",
              freq.cutpoint = 30,
              bins = 25,
              facet = "Antigen") +
  scale_color_viridis_d()
```


## Visualize selected contig sequences on UMAP

```{r, fig.width=10}
contig <- abundanceContig(vdj, cloneCall = "aa", exportTable = TRUE) %>%
  arrange(desc(Abundance)) %>%
  filter(!is.na(CTaa)) %>%
  slice(1:5)
Idents(expression) <- expression$CTaa
DimPlot(expression,
        reduction = "umap",
        cells.highlight = CellsByIdentities(expression, idents = contig$CTaa),
        cols.highlight = turbo(length(contig$CTaa)))
```

## Display V-J combinations on Circos plot

```{r}
norm.COVID.circlize <- expression@meta.data %>%
  filter(Sample_Name_short == "norm.COVID" & !is.na(CTgene)) %>%
  separate(CTgene, into = c("CTgene1", "CTgene2"), sep = "_") %>%
  separate(CTgene1, into = c("v1", "j1", "c1"), sep = "\\.") %>%
  count(v1, j1) %>%
  filter(!is.na(v1) & !is.na(j1) & n > 5) %>%
  arrange(desc(n))

grid.cols <- turbo(length(unique(norm.COVID.circlize$v1)))
names(grid.cols) <- unique(norm.COVID.circlize$v1)
chordDiagram(norm.COVID.circlize, grid.col = grid.cols)
circos.clear()
```

## Session Information
```{r}
sessionInfo()
```
